RSDK-4788: Add Readings to Python SDK

martha-johnston opened the pull request:
fix imports (from sensor to common) and use `GetReadingsRequest/Response` for movement sensor and power sensor

martha-johnston commented on src/viam/components/power_sensor/power_sensor.py:
```diff
@@ -57,6 +57,7 @@ async def get_readings(self, *, extra: Optional[Dict[str, Any]] = None, timeout:
                is_ac: bool
                power: float
             }
+
```
should the below be removed? in `sensor.py` the only thing in `get_readings` is ...

martha-johnston commented on src/viam/components/movement_sensor/movement_sensor.py:
```diff
@@ -147,6 +147,7 @@ async def get_readings(self, *, extra: Optional[Dict[str, Any]] = None, timeout:
                 compass: float,
                 orientation: Orientation,
             }
+
```
should the below be removed? in `sensor.py` the only thing in `get_readings` is `...`. Should we not be calling everything manually now that we have the request/response to use?

benjirewis replied:
@njooma would definitely know. My guess is that now that you're not invoking `super().get_readings()`, you only need `...` here.

randhid replied:
yes we should remove the bespoke implementation in each sdk, we don't need it anymore. It will now be the responsibility of the driver using these wrappers.

njooma replied:
This was modeled after what's done in [rdk](https://github.com/viamrobotics/rdk/blob/12d225bb4da41d580576a44e8047bb5e9e88ce20/components/movementsensor/movementsensor.go#L106) where the base class for Movement Sensor defines the `Readings` function that all subclasses can use.

randhid replied:
@njooma, we added a common.proto `GetReadings` API, so we no longer have to do this in each SDK. Now we have a client and server-side implementation, and we're also changing the [go SDK](https://github.com/viamrobotics/rdk/pull/3047) implementation. So the way it's done now should be changed.

benjirewis commented on src/viam/components/movement_sensor/client.py:
```diff
@@ -3,7 +3,7 @@
 
 from viam.components.movement_sensor.movement_sensor import MovementSensor
-from viam.proto.common import DoCommandRequest, DoCommandResponse, Geometry
+from viam.proto.common import DoCommandRequest, DoCommandResponse, Geometry, GetReadingsRequest, GetReadingsResponse
```
I see some import errors for these values in the test; I'm guessing `GetReadingsRe*` types were added to the `api` repo but they have yet to be automatically imported to the Python SDK.

njooma replied:
Ya exactly -- this has to be merged in after the API changes get merged

martha-johnston commented:
@njooma @benjirewis @clintpurser does anyone know if I should be running make buf locally and uploading that? or should those be generated through GitHub actions? seems like test aren't passing because it doesn't have the new api methods (but it should because api was merged)

benjirewis commented on src/viam/services/navigation/navigation.py:
```diff
@@ -5,6 +5,7 @@
 
 from ..service_base import ServiceBase
 from . import GeoObstacle, GeoPoint, Mode, Waypoint
+from viam.proto.service.navigation import Path
```
@njooma here and elsewhere we were just importing `Path` from `viam.proto.service.navigation`. I see some other API types being imported from `.`; how does that work? We added `Path` (along with `GetPathsRequest` and `GetPathsResponse`) to `src/viam/proto/service/navigation/__init__.py`, but I don't think we were able to find `Path` in `.` here.

njooma replied:
The `.` syntax is only if the class is included in the `src/viam/services/navigation/__init__.py`, not the `.../proto/...` file.

benjirewis commented on tests/test_motion_service.py:
```diff
@@ -17,7 +17,8 @@
 MOTION_CONFIGURATION = MotionConfiguration(
-    vision_services=[ResourceName(namespace="rdk", type="service", subtype="vision", name="viz1")],
+    obstacle_detectors=[ObstacleDetector(vision_service=ResourceName(namespace="rdk", type="service", subtype="vision", name="viz1"),
+                                         camera=ResourceName(namespace="rdk", type="component", subtype="camera", name="cam1"))],
```
I assume it's fine to just use a random camera here in the test?

randhid replied:
Anything that can return `GetObjectPointCluds` for an obstacle. which cameras do.

benjirewis commented on src/viam/services/navigation/client.py:
```diff
@@ -36,6 +39,11 @@ def __init__(self, name: str, channel: Channel):
         self.client = NavigationServiceStub(channel)
         super().__init__(name)
 
+    async def get_paths(self, *, timeout: Optional[float] = None) -> List[Path]:
```
[q] Why do we return API types to users? I thought we would convert the API type to some native Python class and return that to the user?

randhid replied:
I want to request that follow-up API cleanup that doesn't have to do with Martha's readings change be handled in a follow-up ticket. This design does look wrong, though ) I think it should return a Python class but I don't want to make a decision given we're doing this on the fly.

njooma replied:
It's not bad to return a protobuf message. It's actually preferred IMO. This way we don't have to create a bunch of classes that mirror the proto, nor we do we have to keep those classes updated.

benjirewis replied:
@njooma that makes sense. Per recent conversation in the SDKs channel, it seems like it's just language-dependent whether the protos are "good enough" to not need native-mirrored types.

randhid commented:
@njooma @benjirewis @clintpurser  all good?

benjirewis reviewed (approved):
LGTM but probably wait for final approval from @njooma

martha-johnston commented:
@njooma I think I covered everything... I do want movement sensor to no longer be a subclass of sensor so that should be fixed now, I updated the docstring, and added get_readings to ExampleMovementSensor. also updated the nav __init__.py file so path could be imported along with GeoPoint, Waypoint, etc.

njooma commented on src/viam/services/navigation/__init__.py:
```diff
@@ -1,5 +1,5 @@
 from viam.proto.common import GeoObstacle, GeoPoint
-from viam.proto.service.navigation import Mode, Waypoint
+from viam.proto.service.navigation import Mode, Waypoint, Path
```
This will work, but you should also add it in the `__all__` list below

njooma reviewed (approved):
Almost!! One minor change. Approving it because you should be able to merge it in once you add that additional change

martha-johnston commented on tests/test_robot.py:
```diff
@@ -60,7 +60,6 @@
     ResourceName(namespace=RESOURCE_NAMESPACE_RDK, type=RESOURCE_TYPE_COMPONENT, subtype="camera", name="camera1"),
     ResourceName(namespace=RESOURCE_NAMESPACE_RDK, type=RESOURCE_TYPE_COMPONENT, subtype="motor", name="motor1"),
     ResourceName(namespace=RESOURCE_NAMESPACE_RDK, type=RESOURCE_TYPE_COMPONENT, subtype="movement_sensor", name="movement_sensor1"),
-    ResourceName(namespace=RESOURCE_NAMESPACE_RDK, type=RESOURCE_TYPE_COMPONENT, subtype="sensor", name="movement_sensor1"),
```
made the __all__ change, but this was causing the tests to fail... I removed this line because they were both called "movement_sensor1" but this was of subtype "sensor", and now the tests pass. Just want to make sure this checks out @njooma

njooma replied:
Ohh nice find. There's another spot where this should be updated, but I made a [ticket](https://viam.atlassian.net/browse/RSDK-5461) for it for myself for later (not required for this)
